### PR TITLE
Implement io module using Objects

### PR DESCRIPTION
**Pull request by gitlarryf**

### CODE REVIEW

File: exec/cnex/object.h [hunk 1]
```diff
@@ -21,7 +22,9 @@ typedef enum tagNObjectType {
 typedef struct tagTObject {
     ObjectType      type;
     int             refcount;
-    struct tagTCell *cell;
+    void            *cell;
```
Comment (ghewgill) [inline exec/cnex/object.h]: If this isn't going to always be a TCell, then it shouldn't be called `cell`.

File: exec/cnex/lib/io.c [hunk 2]
```diff
@@ -38,7 +64,7 @@ static FILE *check_file(TExecutor *exec, void *pf)
 void io_fprint(TExecutor *exec)
```
Comment (ghewgill) [inline exec/cnex/lib/io.c]: `io.fprint` no longer exists, it's `io.writeLine` now.

File: exec/cnex/lib/io.c [hunk 3]
```diff
@@ -50,7 +76,7 @@ void io_fprint(TExecutor *exec)
 
 void io_close(TExecutor *exec)
 {
-    void *ppf = peek(exec->stack, 0)->address->other;
+    void *ppf = peek(exec->stack, 0)->object->cell;
 
     FILE *f = check_file(exec, ppf);
     fclose(f);
```
Comment (ghewgill) [inline exec/cnex/lib/io.c]: Definitely need to null out the FILE* after closing the file, so you don't try to close it again later.

File: exec/cnex/lib/io.c [hunk 4]
```diff
@@ -65,25 +91,24 @@ void io_open(TExecutor *exec)
 
     const char *m;
     switch (number_to_uint32(mode)) {
-        case ENUM_Mode_read:  m = "r"; break;
-        case ENUM_Mode_write: m = "w+"; break;
+        case ENUM_Mode_read:  m = "rb"; break;
+        case ENUM_Mode_write: m = "wb+"; break;
```
Comment (ghewgill) [inline exec/cnex/lib/io.c]: It's `w+b`, not `wb+`.

File: exec/cnex/object.h [hunk 5]
```diff
@@ -13,6 +13,7 @@ typedef enum tagNObjectType {
     oArray,
     oBoolean,
     oBytes,
+    oFile,
```
Comment (ghewgill) [inline exec/cnex/object.h]: Probably not needed, as discussed.

Comment (gitlarryf) [reply exec/cnex/object.h]: Removed.

File: exec/cnex/lib/io.c (hunk 4 above)
Comment (gitlarryf) [reply exec/cnex/lib/io.c]: Fixed.

File: exec/cnex/lib/io.c (hunk 3 above)
Comment (gitlarryf) [reply exec/cnex/lib/io.c]: Fixed.  Should I also NULL it in release()?

File: exec/cnex/object.h (hunk 1 above)
Comment (gitlarryf) [reply exec/cnex/object.h]: Changed it to `ptr`.  Considered `objPtr`, but decided to go with the shorter name.

File: exec/cnex/lib/io.c (hunk 2 above)
Comment (gitlarryf) [reply exec/cnex/lib/io.c]: To be clear, did you mean `textio.writeLine()`?  There is no exported function writeLine in io.neon/cpp.  I will remove this function from io.c.

Comment (ghewgill) [reply exec/cnex/lib/io.c]: Yes, sorry I did mean `textio.writeLine`. Also, there is no `io.readLine`. That's the whole reason for existence of the `textio` module.

File: exec/cnex/lib/io.c [hunk 6]
```diff
@@ -22,6 +22,31 @@
...
+
+void object_releaseFileObject(Object *o)
+{
+    if (o != NULL) {
+        assert(o->refcount > 0);
+        o->refcount--;
+
+        if (o->refcount <= 0) {
+            if (o->ptr != NULL) {
+                fclose(o->ptr);
```
Comment (ghewgill) [inline exec/cnex/lib/io.c]: You will probably have to cast this to a `FILE*` for at least some compiler.

File: exec/cnex/lib/io.c [hunk 7]
```diff
@@ -35,25 +60,13 @@ static FILE *check_file(TExecutor *exec, void *pf)
...
-}
-
 void io_close(TExecutor *exec)
 {
-    void *ppf = peek(exec->stack, 0)->address->other;
+    void *ppf = peek(exec->stack, 0)->object->ptr;
 
     FILE *f = check_file(exec, ppf);
     fclose(f);
+    f = NULL;
```
Comment (ghewgill) [inline exec/cnex/lib/io.c]: This isn't going to do the trick. You're clearing the local variable `f`, which isn't going to mean much. What you need to clear is the `ptr` member so further file operations can fail cleanly.

File: exec/cnex/object.c [hunk 8]
```diff
@@ -108,8 +109,8 @@ void object_releaseObject(Object *o)
         o->refcount--;
 
         if (o->refcount <= 0) {
-            if (o->cell != NULL) {
-                cell_freeCell(o->cell);
+            if (o->ptr != NULL) {
+                cell_freeCell(o->ptr);
```
Comment (ghewgill) [inline exec/cnex/object.c]: This doesn't look right. What if `ptr` isn't a cell? Can that happen here?

File: exec/cnex/lib/io.c (hunk 2 above)
Comment (gitlarryf) [reply exec/cnex/lib/io.c]: Removed io.readLine() also.

File: exec/cnex/object.c (hunk 8 above)
Comment (gitlarryf) [reply exec/cnex/object.c]: No, it shouldn't EVER happen here.  But it is actually the responsibility of the caller to make sure that the default object_releaseObject() never gets called on any non-cell objects.  I have no way to test to see if the object is a Cell or not either.  Unless I add custom enums for the Objects types.

File: exec/cnex/lib/io.c (hunk 6 above)
Comment (gitlarryf) [reply exec/cnex/lib/io.c]: Fixed.

File: exec/cnex/lib/io.c (hunk 7 above)
Comment (gitlarryf) [reply exec/cnex/lib/io.c]: Fixed.

**Pull request merged**